### PR TITLE
added an initial version of a CMA optimizer

### PR DESCRIPTION
**Pull request by RoyStegeman**
I added a first version of a CMA-ES optimizer in keras as suggested in #16 , but some work is still to be done:
- <s>add comments</s> 
- <s>make it so it only changes trainable layers</s>
- <s>cma decides on number of generations, hence epochs=1. So we should do something to create `history`</s>

The optimizer used is https://pypi.org/project/cma/ by Nikolaus.Hansen, which (according to himself) is supposed to be more efficient than the relatively simple implementation discussed in his papers.

### CODE REVIEW

**Comment by scarlehoff** [comment]
I just tried running this in `n3fit` and seems to work ok (yey!) and the chi2 goes down.
It is much slower but this is expected (and don't really now whether it can be improved, but we can try)

**Comment by RoyStegeman** [comment]
Yes that was to be expected. I have not had a look at Hansen's source code, but CMA is based on matrix manupulation, which I believe python is not too well suited for. Also, the `n3fit` network is somewhat large for using CMA anyway, but there is nothing we can do about that.

**Comment by RoyStegeman** [comment]
@scarlehoff I would like to merge this. Can you have a look?

File: src/evolutionary_keras/models.py [hunk 1]
```diff
@@ -17,6 +17,7 @@
     "ceressolver": Evolutionary_Optimizers.CeresSolver,
 }
 
+
 class EvolModel(Model):
     """
     EvolModel forewards all tasks to keras if the optimizer is NOT genetic.
```
Comment (scarlehoff) [inline src/evolutionary_keras/models.py]: Just notice the typo, forewards -> forwards

File: src/evolutionary_keras/models.py [hunk 2]
```diff
@@ -74,6 +75,10 @@ def perform_genetic_fit(
         for epoch in range(epochs):
             # Generate the best mutant
             score, best_mutant = self.opt_instance.run_step(x=x, y=y)
+            # If the score is a dictionary, it is assumed that the optimizer created a history object
+            if isinstance(score, dict):
```
Comment (scarlehoff) [inline src/evolutionary_keras/models.py]: If verbose == 1 before returning it should print some information.

File: src/evolutionary_keras/optimizers.py [hunk 3]
```diff
@@ -216,7 +217,158 @@ def run_step(self, x, y):
...
+        for i in range(number_to_remove):
+            del self.weightscopy[remove_these_weights[number_to_remove - 1 - i]]
+        self.shape = [weight.shape.as_list() for weight in self.weightscopy]
+        return self.shape
+
+    """ 
+    'weights_per_layer' creates 'self.lengt_flat_layer' which is a list conatining
+    the numer of weights in each layer of the network. 
+    """
+
```
Comment (scarlehoff) [inline src/evolutionary_keras/optimizers.py]: The docstring  should be below the definition of the method.

File: src/evolutionary_keras/optimizers.py [hunk 4]
```diff
@@ -216,7 +217,158 @@ def run_step(self, x, y):
...
+        self.weightscopy = deepcopy(self.model.weights)
+        remove_these_weights = []
+        for i in range(len(self.model.weights)):
+            weight = self.model.weights[i]
+            if weight.name not in self.trainable_weights_names:
+                remove_these_weights.append(i)
+        number_to_remove = len(remove_these_weights)
+        for i in range(number_to_remove):
+            del self.weightscopy[remove_these_weights[number_to_remove - 1 - i]]
+        self.shape = [weight.shape.as_list() for weight in self.weightscopy]
```
Comment (scarlehoff) [inline src/evolutionary_keras/optimizers.py]: I don’t think I understand this logic. Wouldn’t it make more sense to get directly the weights from trainable_weights?

File: src/evolutionary_keras/optimizers.py [hunk 5]
```diff
@@ -216,7 +217,158 @@ def run_step(self, x, y):
...
+        in determining the range of weights in the function 'undo_flatten'. 
+        """
+        self.length_flat_layer = [0]
+        for layer in range(len(self.shape)):
+            flat_layer = np.reshape(self.weightscopy[layer].numpy(), [-1])
+            self.length_flat_layer.append(len(flat_layer))
+
+    """
+    'flatten' returns a 1 dimensional list of all weights in the keras model 
+    """
```
Comment (scarlehoff) [inline src/evolutionary_keras/optimizers.py]: Same thing about the docstring. You can see whether you put the docstring in the correct place by making the documentation (make html in the docs folder) locally because the methods will show up with its docstring,

File: src/evolutionary_keras/optimizers.py [hunk 6]
```diff
@@ -216,7 +217,158 @@ def run_step(self, x, y):
...
+
+    def flatten(self):
+        weights = self.weightscopy
+        flattened_weights = []
+        for layer in range(len(weights)):
+            flattened_weights.append(np.reshape(weights[layer].numpy(), [-1]))
```
Comment (scarlehoff) [inline src/evolutionary_keras/optimizers.py]: Here instead of doing `for layer in range(len(weights))` and then doing `weights[layer]` you can directly do `for weight in weights`

File: src/evolutionary_keras/optimizers.py [hunk 7]
```diff
@@ -216,7 +217,158 @@ def run_step(self, x, y):
...
+    """
+
+    def undo_flatten(self, flattened_weights):
+        new_weights = []
+        for i in range(len(self.shape)):
+            flat_layer = flattened_weights[
+                self.length_flat_layer[i] : self.length_flat_layer[i]
+                + self.length_flat_layer[i + 1]
+            ]
+            new_weights.append(np.reshape(flat_layer, self.shape[i]))
```
Comment (scarlehoff) [inline src/evolutionary_keras/optimizers.py]: Same here, you can do `for i, layer_shape in enumerate(self.shape)` (I would call the array `self.shapes` since there are more than one).

File: src/evolutionary_keras/optimizers.py [hunk 8]
```diff
@@ -216,7 +217,158 @@ def run_step(self, x, y):
...
+        ordered_names = [
+            weight.name for layer in self.model.layers for weight in layer.weights
+        ]
+
+        new_parent = deepcopy(self.model.get_weights())
+        i = 0
+        for weight in self.trainable_weights_names:
+            location_weight = ordered_names.index(weight)
+            new_parent[location_weight] = new_weights[i]
+            i += 1
```
Comment (scarlehoff) [inline src/evolutionary_keras/optimizers.py]: Same, instead of keeping a counter you can use the `enumerate` function: `for i, weight in enumerate(self.trainable_weight_names)`. The `i` will have the same value it has for you there.

File: src/evolutionary_keras/optimizers.py [hunk 9]
```diff
@@ -216,7 +217,158 @@ def run_step(self, x, y):
...
+    epochs, as this is dealth with by 'cma'. The default 'epochs' in EvolModel
+    is one, meaning 'run step' is only called once during training. 
+    """
+
+    def run_step(self, x, y):
+        """ Wrapper to the optimizer"""
+
+        """ 
+        Get the nubmer of weights in each keras layer 
+        """
```
Comment (scarlehoff) [inline src/evolutionary_keras/optimizers.py]: Comments in the middle of the class/methods should go with #, docstrings should go at the beginning of the methods/functions. Strings in the middle to first approximation are just ignored by python.

File: src/evolutionary_keras/optimizers.py [hunk 10]
```diff
@@ -216,7 +217,158 @@ def run_step(self, x, y):
...
+            score = self.model.evaluate(x=x, y=y, verbose=0)
+            loss = parse_eval(score)
+            try:
+                history_data = dict(zip(metricas, score))
+            except TypeError as e:
+                if loss == score:
+                    score = [score, score]
+                    history_data = dict(zip(metricas, score))
+                else:
+                    raise e
```
Comment (scarlehoff) [inline src/evolutionary_keras/optimizers.py]: What is history data used here for? Maybe I’m missing something, but I think you can just return `loss`

File: src/evolutionary_keras/optimizers.py [hunk 11]
```diff
@@ -216,7 +217,158 @@ def run_step(self, x, y):
...
+        xopt, es = cma.fmin2(minimizethis, self.flatten(), self.sigma_init)
+
+        """ 
+        Transform 'xopt' to the models' weight shape. 
+        """
+        selected_parent = self.undo_flatten(xopt)
+
+        """ 
+        Determine the ultimatly selected mutants' performance on the training data. 
+        """
```
Comment (scarlehoff) [inline src/evolutionary_keras/optimizers.py]: Same for these three strings, they should just be comments.

**Comment by RoyStegeman** [comment]
@scarlehoff Thank you for the many comments. I implemented all of them.

**Comment by scarlehoff** [comment]
This is the version you used for the last report in NNPDF right?

I'll change the `setup.py` and the conda recipe and then I'll merge this in if that's ok.

**Comment by RoyStegeman** [comment]
Yes but let me check, I think there are some things that I should adjust for more general use.

**Comment by scarlehoff** [comment]
ok, no rush, we can safely wait until the proceedings are published.

**Comment by RoyStegeman** [comment]
Okay, this should do. Feel free to merge after you made the changes to `setup.py`.

**Comment by RoyStegeman** [comment]
I think this should do, as far as making `evolutionary-keras` future proof is concerned. Pylint does give import errors, but I think that's a problem in pylint. Can you have a look at this error and the setup.py @scarlehoff ?

**Comment by scarlehoff** [comment]
This last part was better suited on a different branch (on a different PR) which can point to this one.

**Comment by RoyStegeman** [comment]
You want me to move them to a separate branch before you have a look?

**Comment by scarlehoff** [comment]
> You want me to move them to a separate branch before you have a look?

Yes please, also because I think we want to wait for Tensorflow 2.2 https://github.com/tensorflow/tensorflow/releases/tag/v2.2.0-rc3 because they have changed the metrics in the `tensorflow.keras` module which can interfere with `evolutionary-keras`.

**Comment by scarlehoff** [comment]
I'll merge this as soon as the tests pass

**Comment by RoyStegeman** [comment]
Sure, thanks for checking the PRs.
